File: src/claude_config/egress_broker/policy.py
"""Egress-broker policy file loader and validator.

Loads per-alias YAML policy files from a directory (typically
``/etc/claude-config/egress-policy/``) and exposes them as a typed
in-memory ``PolicySet`` the request handler can query.

Validation matches ``sandbox/egress-policy/README.md`` exactly. The
loader is transactional: any single invalid file aborts the load
without mutating prior state.
"""

from __future__ import annotations

import fnmatch
import re
from dataclasses import dataclass, field
from pathlib import Path

import yaml

_FQDN_RE = re.compile(
    r"^(?=.{1,253}$)([a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)(?:\.([a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?))+$",
    re.IGNORECASE,
)
_PATH_FORBIDDEN_RE = re.compile(r"[^A-Za-z0-9_\-./*?+%@:,~&=]")
_HARD_CAP_BYTES = 100 * 1024 * 1024
_ALWAYS_STRIPPED_HEADERS = frozenset({"host", "authorization", "cookie", "forwarded"})


class PolicyError(ValueError):
    """A policy file failed validation."""


@dataclass(frozen=True)
class AttachSpec:
    """How a credential is attached to the upstream request."""

    type: str  # header | bearer | query
    name: str  # header name (type=header) or query key (type=query); unused for bearer


@dataclass(frozen=True)
class CredentialSpec:
    """Where to fetch the credential from."""

    backend: str  # pass | stub
    path: str
    attach: AttachSpec


@dataclass(frozen=True)
class UpstreamSpec:
    """Upstream destination the broker forwards to."""

    host: str
    port: int = 443
    scheme: str = "https"


@dataclass(frozen=True)
class Constraints:
    """Request-shape constraints the sandbox must satisfy."""

    methods: frozenset[str]
    paths: tuple[str, ...]
    max_request_bytes: int = 10 * 1024 * 1024
    timeout_seconds: int = 120
    block_request_headers: frozenset[str] = field(default_factory=frozenset)

    def path_allowed(self, path: str) -> bool:
        """Return True if ``path`` matches any allow-listed glob."""
        return any(fnmatch.fnmatchcase(path, pat) for pat in self.paths)

    def stripped_headers(self, attach: AttachSpec) -> frozenset[str]:
        """Lowercase set of header names the broker must strip before forwarding."""
        always = set(_ALWAYS_STRIPPED_HEADERS)
        always.update(h.lower() for h in self.block_request_headers)
        if attach.type == "header":
            always.add(attach.name.lower())
        # X-Forwarded-* family
        return frozenset(always)


@dataclass(frozen=True)
class Policy:
    """A single per-alias policy."""

    alias: str
    upstream: UpstreamSpec
    credential: CredentialSpec
    constraints: Constraints


@dataclass(frozen=True)
class PolicySet:
    """Collection of loaded policies, keyed by alias."""

    by_alias: dict[str, Policy]

    def get(self, alias: str) -> Policy | None:
        """Return the policy for ``alias`` or None if unknown."""
        return self.by_alias.get(alias)


def load_directory(path: Path) -> PolicySet:
    """Load and validate every ``*.yaml`` file under ``path``.

    Args:
        path: Directory containing per-alias YAML files.

    Returns:
        A populated ``PolicySet``.

    Raises:
        PolicyError: If any file fails validation. No partial state is loaded.
        FileNotFoundError: If ``path`` does not exist.
    """
    if not path.is_dir():
        raise FileNotFoundError(f"policy directory does not exist: {path}")

    loaded: dict[str, Policy] = {}
    for yaml_path in sorted(path.glob("*.yaml")):
        policy = _load_file(yaml_path)
        if policy.alias in loaded:
            raise PolicyError(f"duplicate alias '{policy.alias}' at {yaml_path}")
        loaded[policy.alias] = policy
    return PolicySet(by_alias=loaded)


def _load_file(path: Path) -> Policy:
    """Load and validate a single policy file."""
    with path.open("rb") as fp:
        raw = yaml.safe_load(fp)
    if not isinstance(raw, dict):
        raise PolicyError(f"{path}: top-level YAML must be a mapping")

    expected_alias = path.stem
    alias = raw.get("alias")
    if alias != expected_alias:
        raise PolicyError(f"{path}: alias '{alias!r}' must match filename stem '{expected_alias}'")

    upstream = _parse_upstream(path, raw.get("upstream"))
    credential = _parse_credential(path, raw.get("credential"))
    constraints = _parse_constraints(path, raw.get("constraints"), credential.attach)

    return Policy(
        alias=alias,
        upstream=upstream,
        credential=credential,
        constraints=constraints,
    )


def _parse_upstream(path: Path, raw: object) -> UpstreamSpec:
    if not isinstance(raw, dict):
        raise PolicyError(f"{path}: missing or invalid 'upstream' mapping")
    host = raw.get("host")
    if not isinstance(host, str) or not _FQDN_RE.match(host):
        raise PolicyError(f"{path}: 'upstream.host' must be a valid FQDN, got {host!r}")
    scheme = raw.get("scheme", "https")
    if scheme != "https":
        raise PolicyError(f"{path}: 'upstream.scheme' must be 'https', got {scheme!r}")
    port_raw = raw.get("port", 443)
    if not isinstance(port_raw, int) or not (1 <= port_raw <= 65535):
        raise PolicyError(f"{path}: 'upstream.port' must be int 1..65535, got {port_raw!r}")
    return UpstreamSpec(host=host, port=port_raw, scheme=scheme)


def _parse_credential(path: Path, raw: object) -> CredentialSpec:
    if not isinstance(raw, dict):
        raise PolicyError(f"{path}: missing or invalid 'credential' mapping")
    backend = raw.get("backend")
    if backend not in ("pass", "stub"):
        raise PolicyError(f"{path}: 'credential.backend' must be 'pass' or 'stub', got {backend!r}")
    cred_path = raw.get("path")
    if not isinstance(cred_path, str) or not cred_path:
        raise PolicyError(f"{path}: 'credential.path' must be a non-empty string")
    attach_raw = raw.get("attach")
    if not isinstance(attach_raw, dict):
        raise PolicyError(f"{path}: missing or invalid 'credential.attach' mapping")
    attach_type = attach_raw.get("type")
    if attach_type not in ("header", "bearer", "query"):
        raise PolicyError(
            f"{path}: 'credential.attach.type' must be header|bearer|query, got {attach_type!r}"
        )
    attach_name: str
    if attach_type == "bearer":
        attach_name = "Authorization"
    else:
        raw_name = attach_raw.get("name")
        if not isinstance(raw_name, str) or not raw_name:
            raise PolicyError(f"{path}: 'credential.attach.name' required for type={attach_type}")
        attach_name = raw_name
    return CredentialSpec(
        backend=backend,
        path=cred_path,
        attach=AttachSpec(type=attach_type, name=attach_name),
    )


def _parse_constraints(path: Path, raw: object, attach: AttachSpec) -> Constraints:
    if not isinstance(raw, dict):
        raise PolicyError(f"{path}: missing or invalid 'constraints' mapping")
    methods_raw = raw.get("methods")
    if not isinstance(methods_raw, list) or not methods_raw:
        raise PolicyError(f"{path}: 'constraints.methods' must be a non-empty list")
    methods: set[str] = set()
    for m in methods_raw:
        if not isinstance(m, str) or m != m.upper():
            raise PolicyError(f"{path}: method {m!r} must be uppercase ASCII")
        methods.add(m)
    if attach.type == "query" and methods != {"GET"}:
        raise PolicyError(f"{path}: query-string credentials only permitted with methods=[GET]")

    paths_raw = raw.get("paths")
    if not isinstance(paths_raw, list) or not paths_raw:
        raise PolicyError(f"{path}: 'constraints.paths' must be a non-empty list")
    paths: list[str] = []
    for p in paths_raw:
        if not isinstance(p, str) or not p.startswith("/"):
            raise PolicyError(f"{path}: path {p!r} must be a string starting with '/'")
        if ".." in p:
            raise PolicyError(f"{path}: path {p!r} contains forbidden '..'")
        if _PATH_FORBIDDEN_RE.search(p):
            raise PolicyError(f"{path}: path {p!r} contains forbidden characters")
        paths.append(p)

    max_bytes = raw.get("max_request_bytes", 10 * 1024 * 1024)
    if not isinstance(max_bytes, int) or max_bytes <= 0 or max_bytes > _HARD_CAP_BYTES:
        raise PolicyError(
            f"{path}: 'max_request_bytes' must be int in (0, {_HARD_CAP_BYTES}], got {max_bytes!r}"
        )

    timeout = raw.get("timeout_seconds", 120)
    if not isinstance(timeout, int) or timeout <= 0:
        raise PolicyError(f"{path}: 'timeout_seconds' must be positive int, got {timeout!r}")

    block_raw = raw.get("block_request_headers", [])
    if not isinstance(block_raw, list):
        raise PolicyError(f"{path}: 'block_request_headers' must be a list")
    block: set[str] = set()
    for h in block_raw:
        if not isinstance(h, str) or not h:
            raise PolicyError(f"{path}: blocked header {h!r} must be a non-empty string")
        block.add(h.lower())

    return Constraints(
        methods=frozenset(methods),
        paths=tuple(paths),
        max_request_bytes=max_bytes,
        timeout_seconds=timeout,
        block_request_headers=frozenset(block),
    )

File: sandbox/broker/internal/policy/policy.go
// Package policy loads and validates per-alias YAML policy files for
// the egress broker. Parity reference:
// src/claude_config/egress_broker/policy.py and the schema at
// sandbox/egress-policy/README.md.
package policy

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

const (
	HardCapBytes      = 100 * 1024 * 1024
	defaultMaxBytes   = 10 * 1024 * 1024
	defaultTimeoutSec = 120
)

var (
	fqdnRE = regexp.MustCompile(`^(?i)(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)+$`)
	// Path glob character allow-list — matches the Python _PATH_FORBIDDEN_RE inverse.
	pathAllowedCharRE = regexp.MustCompile(`^[A-Za-z0-9_\-./*?+%@:,~&=]+$`)

	alwaysStrippedHeaders = stringSet("host", "authorization", "cookie", "forwarded")
)

// Error is returned by every validation failure in this package.
type Error struct{ Msg string }

func (e *Error) Error() string { return e.Msg }

func policyErr(format string, args ...any) error {
	return &Error{Msg: fmt.Sprintf(format, args...)}
}

// Attach is how a credential is attached to the outbound request.
type Attach struct {
	Type string // header | bearer | query
	Name string
}

// Credential is where the secret comes from.
type Credential struct {
	Backend string // pass | stub
	Path    string
	Attach  Attach
}

// Upstream is the destination the broker forwards to.
type Upstream struct {
	Host   string
	Port   int
	Scheme string
}

// Constraints describe what shapes of request the sandbox may issue.
type Constraints struct {
	Methods             map[string]struct{}
	Paths               []string
	MaxRequestBytes     int
	TimeoutSeconds      int
	BlockRequestHeaders map[string]struct{} // lowercase keys
}

// Policy is a single per-alias policy.
type Policy struct {
	Alias       string
	Upstream    Upstream
	Credential  Credential
	Constraints Constraints
}

// PolicySet holds the loaded policies, keyed by alias.
type PolicySet struct {
	byAlias map[string]*Policy
}

// Get returns the named policy or nil if unknown.
func (s *PolicySet) Get(alias string) *Policy {
	if s == nil {
		return nil
	}
	return s.byAlias[alias]
}

// Len returns the number of loaded aliases.
func (s *PolicySet) Len() int {
	if s == nil {
		return 0
	}
	return len(s.byAlias)
}

// Aliases returns the loaded alias names in sorted order.
func (s *PolicySet) Aliases() []string {
	out := make([]string, 0, len(s.byAlias))
	for k := range s.byAlias {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// PathAllowed reports whether path matches any of the policy's globs.
// Glob semantics match Python's fnmatch.fnmatchcase: '*' matches any
// run of characters INCLUDING '/', '?' matches one character.
func (c *Constraints) PathAllowed(path string) bool {
	for _, pat := range c.Paths {
		if fnmatch(pat, path) {
			return true
		}
	}
	return false
}

func fnmatch(pattern, s string) bool {
	var b strings.Builder
	b.WriteByte('^')
	for _, r := range pattern {
		switch r {
		case '*':
			b.WriteString(".*")
		case '?':
			b.WriteByte('.')
		default:
			b.WriteString(regexp.QuoteMeta(string(r)))
		}
	}
	b.WriteByte('$')
	re, err := regexp.Compile(b.String())
	if err != nil {
		return false
	}
	return re.MatchString(s)
}

// StrippedHeaders returns the lowercase set of headers the broker must
// remove before forwarding (always-stripped + policy-blocked + the
// credential-attach header if attach type is header).
func (c *Constraints) StrippedHeaders(attach Attach) map[string]struct{} {
	out := make(map[string]struct{}, len(alwaysStrippedHeaders)+len(c.BlockRequestHeaders)+1)
	for k := range alwaysStrippedHeaders {
		out[k] = struct{}{}
	}
	for k := range c.BlockRequestHeaders {
		out[k] = struct{}{}
	}
	if attach.Type == "header" {
		out[strings.ToLower(attach.Name)] = struct{}{}
	}
	return out
}

// LoadDirectory validates every *.yaml under dir and returns a PolicySet.
// Transactional: any single failure aborts the load with no partial state.
func LoadDirectory(dir string) (*PolicySet, error) {
	info, err := os.Stat(dir)
	if err != nil {
		return nil, err
	}
	if !info.IsDir() {
		return nil, fmt.Errorf("policy path is not a directory: %s", dir)
	}
	entries, err := filepath.Glob(filepath.Join(dir, "*.yaml"))
	if err != nil {
		return nil, err
	}
	sort.Strings(entries)
	loaded := make(map[string]*Policy)
	for _, path := range entries {
		p, err := loadFile(path)
		if err != nil {
			return nil, err
		}
		if _, dup := loaded[p.Alias]; dup {
			return nil, policyErr("duplicate alias %q at %s", p.Alias, path)
		}
		loaded[p.Alias] = p
	}
	return &PolicySet{byAlias: loaded}, nil
}

type rawAttach struct {
	Type string `yaml:"type"`
	Name string `yaml:"name"`
}

type rawCredential struct {
	Backend string    `yaml:"backend"`
	Path    string    `yaml:"path"`
	Attach  rawAttach `yaml:"attach"`
}

type rawUpstream struct {
	Host   string `yaml:"host"`
	Port   *int   `yaml:"port"`
	Scheme string `yaml:"scheme"`
}

type rawConstraints struct {
	Methods             []string `yaml:"methods"`
	Paths               []string `yaml:"paths"`
	MaxRequestBytes     *int     `yaml:"max_request_bytes"`
	TimeoutSeconds      *int     `yaml:"timeout_seconds"`
	BlockRequestHeaders []string `yaml:"block_request_headers"`
}

type rawPolicy struct {
	Alias       string         `yaml:"alias"`
	Upstream    rawUpstream    `yaml:"upstream"`
	Credential  rawCredential  `yaml:"credential"`
	Constraints rawConstraints `yaml:"constraints"`
}

func loadFile(path string) (*Policy, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var raw rawPolicy
	dec := yaml.NewDecoder(strings.NewReader(string(data)))
	dec.KnownFields(false)
	if err := dec.Decode(&raw); err != nil {
		return nil, policyErr("%s: invalid YAML: %s", path, err)
	}

	expectedAlias := strings.TrimSuffix(filepath.Base(path), ".yaml")
	if raw.Alias != expectedAlias {
		return nil, policyErr("%s: alias %q must match filename stem %q", path, raw.Alias, expectedAlias)
	}

	upstream, err := parseUpstream(path, raw.Upstream)
	if err != nil {
		return nil, err
	}
	cred, err := parseCredential(path, raw.Credential)
	if err != nil {
		return nil, err
	}
	cons, err := parseConstraints(path, raw.Constraints, cred.Attach)
	if err != nil {
		return nil, err
	}
	return &Policy{
		Alias:       raw.Alias,
		Upstream:    upstream,
		Credential:  cred,
		Constraints: cons,
	}, nil
}

func parseUpstream(path string, r rawUpstream) (Upstream, error) {
	if r.Host == "" || !fqdnRE.MatchString(r.Host) {
		return Upstream{}, policyErr("%s: 'upstream.host' must be a valid FQDN, got %q", path, r.Host)
	}
	scheme := r.Scheme
	if scheme == "" {
		scheme = "https"
	}
	if scheme != "https" {
		return Upstream{}, policyErr("%s: 'upstream.scheme' must be 'https', got %q", path, scheme)
	}
	port := 443
	if r.Port != nil {
		port = *r.Port
	}
	if port < 1 || port > 65535 {
		return Upstream{}, policyErr("%s: 'upstream.port' must be int 1..65535, got %d", path, port)
	}
	return Upstream{Host: r.Host, Port: port, Scheme: scheme}, nil
}

func parseCredential(path string, r rawCredential) (Credential, error) {
	if r.Backend != "pass" && r.Backend != "stub" {
		return Credential{}, policyErr("%s: 'credential.backend' must be 'pass' or 'stub', got %q", path, r.Backend)
	}
	if r.Path == "" {
		return Credential{}, policyErr("%s: 'credential.path' must be a non-empty string", path)
	}
	t := r.Attach.Type
	if t != "header" && t != "bearer" && t != "query" {
		return Credential{}, policyErr("%s: 'credential.attach.type' must be header|bearer|query, got %q", path, t)
	}
	name := r.Attach.Name
	if t == "bearer" {
		name = "Authorization"
	} else if name == "" {
		return Credential{}, policyErr("%s: 'credential.attach.name' required for type=%s", path, t)
	}
	return Credential{
		Backend: r.Backend,
		Path:    r.Path,
		Attach:  Attach{Type: t, Name: name},
	}, nil
}

func parseConstraints(path string, r rawConstraints, attach Attach) (Constraints, error) {
	if len(r.Methods) == 0 {
		return Constraints{}, policyErr("%s: 'constraints.methods' must be a non-empty list", path)
	}
	methods := make(map[string]struct{}, len(r.Methods))
	for _, m := range r.Methods {
		if m == "" || m != strings.ToUpper(m) {
			return Constraints{}, policyErr("%s: method %q must be uppercase ASCII", path, m)
		}
		methods[m] = struct{}{}
	}
	if attach.Type == "query" {
		if _, ok := methods["GET"]; !ok || len(methods) != 1 {
			return Constraints{}, policyErr("%s: query-string credentials only permitted with methods=[GET]", path)
		}
	}
	if len(r.Paths) == 0 {
		return Constraints{}, policyErr("%s: 'constraints.paths' must be a non-empty list", path)
	}
	paths := make([]string, 0, len(r.Paths))
	for _, p := range r.Paths {
		if !strings.HasPrefix(p, "/") {
			return Constraints{}, policyErr("%s: path %q must be a string starting with '/'", path, p)
		}
		if strings.Contains(p, "..") {
			return Constraints{}, policyErr("%s: path %q contains forbidden '..'", path, p)
		}
		if !pathAllowedCharRE.MatchString(p) {
			return Constraints{}, policyErr("%s: path %q contains forbidden characters", path, p)
		}
		paths = append(paths, p)
	}
	maxBytes := defaultMaxBytes
	if r.MaxRequestBytes != nil {
		maxBytes = *r.MaxRequestBytes
	}
	if maxBytes <= 0 || maxBytes > HardCapBytes {
		return Constraints{}, policyErr("%s: 'max_request_bytes' must be int in (0, %d], got %d", path, HardCapBytes, maxBytes)
	}
	timeout := defaultTimeoutSec
	if r.TimeoutSeconds != nil {
		timeout = *r.TimeoutSeconds
	}
	if timeout <= 0 {
		return Constraints{}, policyErr("%s: 'timeout_seconds' must be positive int, got %d", path, timeout)
	}
	block := make(map[string]struct{}, len(r.BlockRequestHeaders))
	for _, h := range r.BlockRequestHeaders {
		if h == "" {
			return Constraints{}, policyErr("%s: blocked header must be a non-empty string", path)
		}
		block[strings.ToLower(h)] = struct{}{}
	}
	return Constraints{
		Methods:             methods,
		Paths:               paths,
		MaxRequestBytes:     maxBytes,
		TimeoutSeconds:      timeout,
		BlockRequestHeaders: block,
	}, nil
}

func stringSet(items ...string) map[string]struct{} {
	out := make(map[string]struct{}, len(items))
	for _, x := range items {
		out[x] = struct{}{}
	}
	return out
}
